bindings/rust/s2n-tls/src/testing/s2n_tls.rs

```diff
@@ -0,0 +1,325 @@
+
+        if remaining_capacity < data.len() {
+            // Flush the send buffer before reallocating it
+            self.flush();
+
+            // ensure we only do one allocation for this write
+            let len = SEND_BUFFER_CAPACITY.max(data.len());
+
+            debug_assert!(
+                self.send_buffer.is_empty(),
+                "dropping a send buffer with data will result in data loss"
+            );
+            *self.send_buffer = BytesMut::with_capacity(len);
+        }
+
+        // Write the current data to the send buffer
+        //
+        // NOTE: we don't immediately flush to the context since s2n-tls may do
+        //       several small writes in a row.
+        self.send_buffer.extend_from_slice(data);
+
+        data.len()
+    }
+
+    /// Flushes the send buffer into the context
+    fn flush(&mut self) {
+        if !self.send_buffer.is_empty() {
+            let chunk = self.send_buffer.split().freeze();
+            self.context.send(chunk);
+        }
+    }
+
+    /// The function s2n-tls calls when it wants to receive data
+    unsafe extern "C" fn recv_cb(context: *mut c_void, data: *mut u8, len: u32) -> s2n_status_code {
+        let context = &mut *(context as *mut Self);
+        let data = core::slice::from_raw_parts_mut(data, len as _);
+        match context.on_read(data) {
+            0 => {
+                // https://github.com/awslabs/s2n/blob/main/docs/USAGE-GUIDE.md#s2n_connection_set_send_cb
+                // s2n-tls wants us to set the global errno to signal blocked
+                errno::set_errno(errno::Errno(libc::EWOULDBLOCK));
+                -1
+            }
+            len => len as _,
+        }
+    }
+
+    /// Called when receiving data
+    fn on_read(&mut self, data: &mut [u8]) -> usize {
+        let max_len = Some(data.len());
+
+        if let Some(chunk) = self.context.receive(max_len) {
+            let len = chunk.len();
+            //dbg!(len);
+            data[..len].copy_from_slice(&chunk);
+            len
+        } else {
+            0
+        }
+    }
+}
+
+pub mod tests {
+    use super::*;
+    use crate::raw::{config::*, error::Error};
+    use crate::testing::Pair;
+
+    const SAMPLES: usize = 100;
+
+    #[allow(unused_variables)]
+    struct HostVerifyData {
+        callback_invoked: bool,
+        allow: bool,
+        host: &'static str,
+    }
+    struct CertKeyPair {
+        cert: &'static [u8],
+        key: &'static [u8],
+    }
+    #[allow(unused_variables, dead_code)]
+    impl CertKeyPair {
+        fn default() -> Self {
+            CertKeyPair {
+                cert: include_str!("../../../../../tests/pems/rsa_4096_sha512_client_cert.pem")
+                    .as_bytes(),
+                key: include_str!("../../../../../tests/pems/rsa_4096_sha512_client_key.pem")
+                    .as_bytes(),
+            }
+        }
+        fn ecdsa() -> Self {
```

Comment:
Spaces between functions, please.

Reply:
sorry- on the hunt for a better IDE formatter 😞